Shared/Realm.Sync.Shared/Credentials.cs

```diff
@@ -0,0 +1,162 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Json;
+
+namespace Realms.Sync
+{
+    public class Credentials
+    {
+        internal static class Providers
+        {
+            internal const string Debug = "debug";
+
+            internal const string Facebook = "facebook";
+
+            internal const string Google = "google";
+
+            internal const string Twitter = "twitter";
+
+            internal const string Password = "password";
+
+            internal const string AccessToken = "accessToken";
+        }
+
+        internal static class Keys
+        {
+            internal const string CreateUser = "register";
+
+            internal const string Password = "password";
+
+            internal const string Identity = "identity";
+
+            internal const string IsAdmin = "isAdmin";
+        }
+
+        public static Credentials Custom(string identityProvider, string userIdentifier,
+            IReadOnlyDictionary<string, object> userInfo)
```

Comment:
You can still pass a `Dictionary<string, object>` because it, strangely enough, implements the interface.